pkg/tanka/list.go

```diff
@@ -0,0 +1,101 @@
+package tanka
+
+import (
+	"io/ioutil"
+	"path/filepath"
+
+	"github.com/grafana/tanka/pkg/jsonnet"
+	"github.com/grafana/tanka/pkg/spec/v1alpha1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+// ListOpts are optional arguments for ListEnvs
+type ListOpts struct {
+	Selector labels.Selector
+}
+
+// ListEnvs returns metadata of all environments recursively found in 'dir'.
+// Each directory is tested and included if it is a valid environment, either
+// static or inline. If a directory is a valid environment, its subdirectories
+// are not checked.
+func ListEnvs(dir string, opts ListOpts) ([]*v1alpha1.Environment, error) {
+	// list all environments at dir
+	envs, err := list(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	// optionally filter
+	if opts.Selector == nil || opts.Selector.Empty() {
+		return envs, nil
+	}
+
+	filtered := make([]*v1alpha1.Environment, 0, len(envs))
+	for _, e := range envs {
+		if !opts.Selector.Matches(e.Metadata) {
+			continue
+		}
+		filtered = append(filtered, e)
+	}
+
+	return filtered, nil
+}
+
+// list implements the actual functionality described at 'ListEnvs'
+func list(dir string) ([]*v1alpha1.Environment, error) {
```

Comment:
This is essentially `ParseParallel` except it completely takes away to use `EvalScript` in combination with parallel processing of environments. This has a performance impact on the export functionality, which I mistakenly allowed to be removed in PR #459 as I thought I could bring it back if we modify the Loaders interface little bit.

`Peek` in itself is a nice interface but it is has only one dumb function instead of being able to leverage opts.JsonnetOpts with EvalScript. Based on this, I actually want to reject both #467 and this PR as it adds more constraints than we benefit from it.

Reply:
I left a comment in #467 that might be able to address the last part.

Reply:
> This is essentially ParseParallel

This is not correct. I wrote a new algorithm here that does file-tree traversal based on the result of individual env parse results. ParseParallel solves a different problem, especially because it requires a pre-computed list of environments.

> one dumb function instead of being able to leverage opts.JsonnetOpts with EvalScript

There is value in simplicity. The `Peek` API gets us very far. Having `EvalScript` is very similar to injecting heaps of logic using closures, etc., which is very hard to understand because you downstream functions behavior changes based on the callers context.

Reply:
> mistakenly allowed to be removed in PR #459 as I thought I could bring it back if we modify the Loaders interface little bit

PR https://github.com/grafana/tanka/pull/465 shows that this is perfectly possible using the `Peek` and `Load` API